### DISCUSSION TITLE
add_font fails second time generate PDF is called - do I need to dispose?

### DISCUSSION THREAD

**Discussion by gumbald** [py-pdf/fpdf2#306] [Q&A]
Hello - I'm adding some fonts from a local directory and using them successfully in a document. However, when I try to run this immediately after, it throws an exception on adding the font again (file not found) - I suspect because the files are still locked. Is there a method I should be using to clean up and unlock the font file...?

**Comment by Lucas-C** [comment]
Hi @gumbald !
No locks are put on font files by `fpdf2`.
Could you please provide some more details so we can help you?
What is the full exception you are seing?
Some [minimal Python code](https://stackoverflow.com/help/minimal-reproducible-example) reproducing your issue would also help a lot.

**Reply by gumbald** [reply]
OK - I've gone back to very basics - and I can't repeat it with a simple example!

In essence, this is what I'm doing, but the second time gives the file not found on the .ttf file...

```
from fpdf import FPDF
import locale
from datetime import datetime
import time

class ReportPDF(FPDF):
    def LocaleTime(self, formatter, time=None):
        if time is None: time = datetime.now()
        locale.setlocale(locale.LC_ALL, 'en')
        return time.strftime(formatter)
    def GenerateReport(self, location):
        self.add_font('TDY', fname='fonts/DejaVuSansCondensed.ttf', uni=True)
        self.add_font('TDY-B', fname='fonts/DejaVuSansCondensed-Bold.ttf', uni=True)
        self.add_font('TDY-I', fname='fonts/DejaVuSansCondensed-Oblique.ttf', uni=True)
        self.InitialiseReport()
        self.SaveReport(location)
    def InitialiseReport(self):
        self.add_page()
        self.set_font('TDY-B', '', 16)
        self.set_xy(0.0,110)
        self.cell(w=210.0, h=40.0, align='C', txt="Testing", border=0)
        self.set_xy(0.0,120)
        self.set_font('TDY-I', '', 14)
        self.set_xy(0.0,150)
        self.cell(w=210.0, h=40.0, align='C', txt=self.LocaleTime("%d %b %Y", datetime.now()), border=0)
    def SaveReport(self, location):
        self.output(location, 'F')

for i in range(4):
    pdf = ReportPDF(orientation='P', unit='mm', format='A4')
    pdf.GenerateReport(r"C:\\Users\\gumbald\\Desktop\\" + datetime.now().strftime("%Y%m%d%H%M%S") + str("_Output.pdf"))
    time.sleep(2)
```

**Reply by gumbald** [reply]
Here's the full exception from my other code:

```
[E 211130 14:24:21 gui:412] Cannot save current data in file 'C:\Users\gumbald\Desktop\20211130142420_Report.pdf'.
    Traceback (most recent call last):
      File "c:\Users\gumbald\source\repos\Badger\gui.py", line 409, in GenerateReport
        pdf.GenerateReport(gvars.results, fn)
      File "c:\Users\gumbald\source\repos\Badger\report.py", line 20, in GenerateReport
        self.add_font('TDY', fname='fonts/DejaVuSansCondensed.ttf', uni=True)
      File "c:\Users\gumbald\source\repos\Badger\.badger\lib\site-packages\fpdf\fpdf.py", line 1105, in add_font
        raise FileNotFoundError(f"TTF Font file not found: {fname}")
    FileNotFoundError: TTF Font file not found: fonts/DejaVuSansCondensed.ttf
```

**Reply by Lucas-C** [reply]
I tested your code on my computer, and it works perfectly fine after downloading the `.ttf` files from: https://github.com/mps/fonts

Maybe you should check that you have a `DejaVuSansCondensed.ttf` file with this exact name in your `fonts/` directory?

**Reply by gumbald** [reply]
It's definitely there - I'm triggering the method using a button, which works first time but not the second. This is the only thing using the .ttf file, but I'm going to have to monitor if something else is having a side effect. Thanks for the responses!

**Comment by gmischler** [comment]
@gumbald , if the file was locked, you'd probably get a PermissionError, but certainly not a FileNotFoundError.

I see that you're using relative paths to the font files. This makes me suspect that your code changes the current directory of the process at some point. After that, the relative paths won't point to the right place anymore.
Try eg. `print(os.getcwd())` at the beginning and end of the relevant routine to see what's going on.

**Reply by gumbald** [reply]
That's it! It seems to (occasionally) change the working directory to the last place I saved a report... Now to track down why! I guess this is a more general Python question - I figured the relative path would be to the execution. How should I reference the font files instead? Different users will have different absolute directories obviously.

EDIT: My colleague can't repeat this running the same version of code...! I also can't repeat it if I put `print(os.getcwd())` at either side of the FPDF `output(location, 'F')` function.

**Reply by gmischler** [reply]
>  I figured the relative path would be to the execution. How should I reference the font files instead?

That depends on how you have your fonts organized.
If I didn't miss anything, then apparently fpdf2 doesn't look in the system fonts directory on its own (it really should!).
It does look in the current directory, and in a `font` subdirectory of wherever it is installed itself (the default value of `fpdf.FPDF_FONT_DIR`).

> EDIT: My colleague can't repeat this running the same version of code...! I also can't repeat it if I put `print(os.getcwd())` at either side of the FPDF `output(location, 'F')` function.

The `output(location)` will not change the current directory (a library should really never do this).
You'll have to check all of your real production code, and anything else that might get executed in between invocations.
